gemoji/src/main/kotlin/io/sweers/catchup/gemoji/AssetSQLiteOpenHelper.kt

```diff
@@ -16,92 +15,113 @@
  * limitations under the License.
  */
 
-// Need package to access FrameworkSQLiteDatabase
-@file:Suppress("PackageDirectoryMismatch")
-package android.arch.persistence.db.framework
+package io.sweers.catchup.gemoji
 
-import android.arch.persistence.db.SupportSQLiteDatabase
-import android.arch.persistence.db.SupportSQLiteOpenHelper
 import android.content.Context
 import android.database.sqlite.SQLiteDatabase
-import com.readystatesoftware.sqliteasset.SQLiteAssetHelper
+import android.database.sqlite.SQLiteOpenHelper
+import java.io.File
+import java.io.FileOutputStream
 
-internal class AssetSQLiteOpenHelper(context: Context, name: String, version: Int,
-    callback: SupportSQLiteOpenHelper.Callback): SupportSQLiteOpenHelper {
-
-  val assetHelper = AssetHelper(context, name, version, callback)
-
-  override fun getDatabaseName(): String {
-    return assetHelper.databaseName
-  }
-
-  override fun setWriteAheadLoggingEnabled(enabled: Boolean) {
-    assetHelper.setWriteAheadLoggingEnabled(enabled)
-  }
-
-  override fun getWritableDatabase(): SupportSQLiteDatabase {
-    return assetHelper.writableSupportDatabase
-  }
-
-  override fun getReadableDatabase(): SupportSQLiteDatabase {
-    return assetHelper.readableSupportDatabase
-  }
+/**
+ * A minimal implementation of an [SQLiteOpenHelper] that is backed by an sqlite db from
+ * `/assets/databases/`.
+ *
+ * This implementation does not support proper migrations. Instead, when a new version
+ * is detected the database from `/assets/databases/` is re-copied into the internal database
+ * directory.
+ *
+ * @param context the context.
+ * @param name the db name, the db should be placed under `/assets/databases/name`.
+ * @param version the db version, when a new version is detected, the asset db is re-copied into
+ * the internal database directory.
+ */
+abstract class AssetSQLiteOpenHelper(private val context: Context,
+    name: String, private val version: Int): SQLiteOpenHelper(context, name, null, version) {
 
-  override fun close() {
-    assetHelper.close()
-  }
-}
+  private val databasePath = "${context.applicationInfo.dataDir}/databases/$name"
+  private val assetPath = "databases/$name"
 
-internal class AssetHelper(context: Context, name: String, private val version: Int,
-    private val callback: SupportSQLiteOpenHelper.Callback):
-    SQLiteAssetHelper(context, name, null, null, version, null) {
+  private var database: SQLiteDatabase? = null
 
-  private var wrappedDb: FrameworkSQLiteDatabase? = null
-  private var updateIdentity = false
+  @Synchronized override fun getWritableDatabase(): SQLiteDatabase {
+    database?.let {
+      if (it.isOpen && !it.isReadOnly) {
+        return@getWritableDatabase it
+      }
+    }
 
-  init {
-    setForcedUpgrade()
+    var success = false
+    var db: SQLiteDatabase? = null
```

Comment:
Any reason not to use `use` here?

Reply:
We only close the `db` if we were unsuccessful for some reason. The other close is on the `database` field. Tried to improve the readability with https://github.com/hzsweers/CatchUp/pull/96/commits/5f39faa99b1b44cfcd8185808a589517622eb559